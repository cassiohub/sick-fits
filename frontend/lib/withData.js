import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { devEndpoint, prodEndpoint } from '../config';

import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? devEndpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart: (_, variables, state) => {
            const { cache } = state;
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });

            const newState = {
              data: { cartOpen: !cartOpen, },
            };

            cache.writeData(newState);

            return newState;
          }
        }
      },
      defaults: {
        cartOpen: false,
      },
    }
  });
}

export default withApollo(createClient);
