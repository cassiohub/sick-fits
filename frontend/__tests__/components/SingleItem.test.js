import wait from 'waait';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';

import { fakeItem } from '../../lib/testUtils';
import SingleItem, { SINGLE_ITEM_QUERY } from '../../components/SingleItem';

describe('Component', () => {
  describe('<SingleItem />', () => {
    it('renders with proper data', async () => {
      const mocks = [
        {
          request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' }},
          result: { data: { item: fakeItem() }},
        },
      ];

      const wrapper = mount(
        <MockedProvider mocks={mocks}>
          <SingleItem id={'123'} />
        </MockedProvider>
      );

      expect(wrapper.text()).toContain('Loading...');

      await wait();
      wrapper.update();

      expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
      expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
      expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
    });

    it('renders error message', async () => {
      const mocks = [
        {
          request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' }},
          result: {
            errors: [
              { message: 'Item not found' }
            ]
          },
        },
      ];

      const wrapper = mount(
        <MockedProvider mocks={mocks}>
          <SingleItem id={'123'} />
        </MockedProvider>
      );

      await wait();
      wrapper.update();

      const error = wrapper.find('[data-test="graphql-error"]');;
      expect(error.text()).toContain('Item not found');
      expect(toJSON(error)).toMatchSnapshot();
    });
  });
});