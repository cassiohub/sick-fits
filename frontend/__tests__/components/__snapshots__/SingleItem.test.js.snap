// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <SingleItem /> renders error message 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Item not found
</p>
`;

exports[`Component <SingleItem /> renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`Component <SingleItem /> renders with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`Component <SingleItem /> renders with proper data 3`] = `
<p>
  dogs
</p>
`;
